### ISSUE TITLE
401 when using remote tunnel

### ISSUE THREAD

**Issue by jayenashar** [microsoft/vscode-livepreview#462]
![image](https://user-images.githubusercontent.com/1137152/222299262-8fe8dd49-4f0b-4d17-8a53-9caea1166a4e.png)

Using http://127.0.0.1:3000/index.html works fine, but requires the remote machine to be 127.0.0.1

Using https://90cmx3jb-3000.aue.devtunnels.ms/index.html returns a 401.

**Comment by jayenashar** [comment]
seems to require the host header to be 127.0.0.1.  i used `mitmproxy --listen-port 8000 --mode reverse:http://127.0.0.1:3000` and setup a port tunnel for port 8000 and it works good.

**Comment by andreamah** [comment]
Related: #381

**Comment by jayenashar** [comment]
:(  even `localhost` doesn't work

**Comment by shirakaba** [comment]
Thanks @jayenashar for the workaround!

Another way to make the reverse proxy (if you don't have `mitmproxy` but do have `ngrok`) whilst adding the expected permissive host header is:

```sh
ngrok http --request-header-add 'host:localhost' 3000
```

For example, if your desktop device is hosting the live preview server on `http://127.0.0.1:3000`, ngrok would create a globally-accessible forwarding address for it:

```
# (Example randomly-generated forwarding address)
https://0b95-126-126-99-39.ngrok-free.app -> http://localhost:3000
```

External devices (e.g. your mobile phone) can then connect to that forwarding address even if they're forbidden from connecting directly to the live preview server.

**Comment by jayenashar** [comment]
the issue isn't that it's listening only on `127.0.0.1`.  the issue is that it requires the HTTP `Host` header to be `127.0.0.1`.  It doesn't even accept `localhost`.

**Comment by jayenashar** [comment]
thanks.  i see the code in https://github.com/microsoft/vscode-livepreview/blob/28f4a25278d8840f6138483f21148d8bd66089eb/src/server/httpServer.ts#L172-L181 .  i assume it's for security and can't be simply removed.  maybe if there is a way to add a whitelist of valid host & origin headers in the plugin settings, that's probably the best fix and may be accepted in a pull request.

**Comment by jayenashar** [comment]
yes, that is my opinion.

i don't know much about writing vscode extensions so i can't really be of much help.  but if you can modify the extension locally, i assume there is a way to install your modified version on the remote.

**Comment by andreamah** [comment]
> thanks. i see the code in
> 
> https://github.com/microsoft/vscode-livepreview/blob/28f4a25278d8840f6138483f21148d8bd66089eb/src/server/httpServer.ts#L172-L181
> 
> . i assume it's for security and can't be simply removed. maybe if there is a way to add a whitelist of valid host & origin headers in the plugin settings, that's probably the best fix and may be accepted in a pull request.

I haven't looked too much into this yet, but is there a specific IP address that we can resolve and whitelist?

**Comment by jayenashar** [comment]
> I haven't looked too much into this yet, but is there a specific IP address that we can resolve and whitelist?

@andreamah i didn't understand your question.  do you mean like 127.0.0.1?

> Wild guess, if there is an active tunnel I guess one can fetch the information from this spesific connection and bypass (?).

i don't understand what you are suggesting.  if there is a tunnel, how can the live preview extension know about it?  probably easier for the whitelist to allow regular expressions like `.*\.devtunnels\.ms`

**Comment by andreamah** [comment]
> i didn't understand your question. do you mean like 127.0.0.1?

Yeah, I just mean how to get the host address that we can use to access the page. I'm assuming that the port is forwarded from the tunnelling machine to the local machine for access somehow, so getting that would help. I can also look into this when I have time, since I haven't took a look at what actually happens in a while.

**Comment by jayenashar** [comment]
the remote tunnels extension is a different extension.  i'm not sure if you can get data from it, but i wouldn't do it that way.  instead, i might create a similar interface in this extension that lists the Host/Origin headers, and has a quick button to whitelist those values/hostnames.

that's a bit more work though, so i think start simple with a whitelist that a user can set in the extension settings.  default to `127.0.0.1` and `localhost`

**Comment by jayenashar** [comment]
@thorgrimjansrud i've never seen that.  have a look at my screenshot at the start of this issue.  mine has only ever shown 127.0.0.1

**Comment by andreamah** [comment]
Yes. Port forwarding is implemented in the core editor, so we can also use the vscode API to resolve the IP or URL and port number that you can use to access it locally. Ideally, we have something that looks like
1. tunnels starts the service on their side and there's a forwarded port and way to access the address locally.
2. live preview gets that address and whitelists it. I *think* I tried to resolve the external host already when looking at what to whitelist, but I'll need to take another look; perhaps it's a timing issue or I overlooked what happens when there's a custom URL like that.

**Comment by jayenashar** [comment]
@thorgrimjansrud you can see in my screenshot i've got the same things open but it still doesn't list the devtunnels.ms in the debug logging.  My setup is a https://vscode.dev/ frontend with a vscode server running on linux.

Even if the external uri method does work, I have the same issue without an external uri.  Just open vscode on my desktop, start live preview, and `localhost` is not accepted.  No remote, no port forward.

**Comment by metacop** [comment]
i have same problem. make a rest api. and when I consult it with postman it returns 401
![image](https://github.com/microsoft/vscode-livepreview/assets/56640826/0be17032-4324-4c99-82b7-83653088ef2c)

**Comment by Emmanuel-Edidiong** [comment]
Hi @metacop, I'm having the same problem for rest api calls. Were you able to find a way around it?

**Comment by metacop** [comment]
> Hi @metacop, I'm having the same problem for rest api calls. Were you able to find a way around it?

No friend, I won't be able to solve it. use ngrok https://ngrok.com/ just download it and run the command ngrok http 80

and clearly pull up the app locally with port 80 to match ngrok. That's right, ngrok is a little slow, but for development it is very good.

**Comment by Emmanuel-Edidiong** [comment]
Okay sadly, thanks. But I somehow found a way around it. When you change the port's 
 visibility from private to public it works and won't return 401. 
 
![screenshot-docs github com-2023 09 21-10_49_16](https://github.com/microsoft/vscode-livepreview/assets/40518930/35b20e3d-68d1-4221-a645-136e7b38935b)
 
 However, it still times out and returns no response

**Comment by Emmanuel-Edidiong** [comment]
> > Hi @metacop, I'm having the same problem for rest api calls. Were you able to find a way around it?
> 
> No friend, I won't be able to solve it. use ngrok https://ngrok.com/ just download it and run the command ngrok http 80
> 
> and clearly pull up the app locally with port 80 to match ngrok. That's right, ngrok is a little slow, but for development it is very good.

Seems like ngrok will still do for now

**Comment by kaif91** [comment]
Thanks @Emmanuel-Edidiong  For me changing the port visibility works with vscode dev tunnel.

**Comment by javaskript** [comment]
I was getting this exact same issue ever since switching to ethernet from WiFi. Here's how I fixed it.

1. In your PC's network settings, make sure your network is set to Private
2. Open cmd and run ipconfig, note your IPv4 address
3. In Live Preview extension settings, set Host IP to your IPv4 address
4. Right click your file in the file tree, Show Preview

That's it, no GitHub auth or port forwarding needed. Just go to "[YourIPv4Address]:3000" on your current or other devices and there will be no more 401 Unauthorized errors.
